jolly-jellyfish/src/page_maker/views.py

```diff
@@ -1,3 +1,219 @@
+
+    def form_valid(self, form):
+        form.save()
+        return super().form_valid(form)
+
+
+class UserDetailView(DetailView):
+    model = User
+    context_object_name = 'viewed_user'
+    template_name = 'page_maker/user_detail.html'
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        # user = get_object_or_404(User, pk=self.kwargs.get('pk'))
+        user = self.get_object()
+        context['user_pages'] = Webpage.objects.filter(author=user)
+        return context
+
+
+class UserUpdateView(LoginRequiredMixin, UserPassesTestMixin, UpdateView):
+    model = User
+    template_name = 'page_maker/user_update.html'
+    fields = ['first_name', 'last_name', 'email']
+    success_url = '/'
+
+    def test_func(self):
+        user = self.get_object()
+        if self.request.user == user:
+            return True
+        return False
+
+
+class UserDeleteView(LoginRequiredMixin, UserPassesTestMixin, DeleteView):
+    model = User
+    template_name = 'page_maker/user_delete.html'
+    success_url = '/'
+
+    def test_func(self):
+        user = self.get_object()
+        if self.request.user == user:
+            return True
+        return False
+
+
+class WebpageCreateView(LoginRequiredMixin, FormView):
+    template_name = 'page_maker/webpage_create.html'
+    form_class = WebpageForm
+
+    # TODO validate and sanitize text and images
+    def form_valid(self, form):
+        self.form = form
+        form.instance.author = self.request.user
+        form.save()
+        return super().form_valid(form)
+
+    def get_success_url(self):
+        pagename = self.form.cleaned_data['name']
+        return reverse_lazy('webpage-view', kwargs={'pagename': pagename})
+
+
+class WebpageView(DetailView):
+    """
+    Display webpage
+    """
+    model = Webpage
+    template_name = 'page_maker/webpage_view.html'
+    context_object_name = 'webpage'
+    slug_field = 'name'
+    slug_url_kwarg = 'pagename'
+
+
+class WebpageDetailView(DetailView):
+    """
+    Display webpage details and comment section
+    """
+    model = Webpage
+    template_name = 'page_maker/webpage_detail.html'
+    context_object_name = 'webpage'
+    slug_field = 'name'
+    slug_url_kwarg = 'pagename'
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        webpage = self.get_object()
+        context['comments'] = Comment.objects.filter(parent_page=webpage)
+        context['comment_form'] = CommentForm()
+        return context
+
+
+class WebpageListView(ListView):
+    """
+    Display webpages ordered by vote number
+    """
+    model = Webpage
+    template = 'page_maker/webpage_list.html'
+    context_object_name = 'webpages'
+    paginate_by = 10
+    ordering = ['-votes']
+
+
+class WebpageUpdateView(LoginRequiredMixin, UserPassesTestMixin, UpdateView):
+    model = Webpage
+    template_name = 'page_maker/webpage_update.html'
```

Comment:
PATHLIB

Reply:
I think this has to be a string actually and is how Django sets it out themselves too (https://docs.djangoproject.com/en/3.0/topics/class-based-views/generic-display/#generic-views-of-objects). It's not actually used as a path I think but as sort of a key to finding the correct template using Django's inner magic 👍 .